src/librbd/image/CreateRequest.cc

```diff
@@ -133,16 +139,94 @@ CreateRequest<I>::CreateRequest(IoCtx &ioctx, const std::string &image_name,
   m_id_obj = util::id_obj_name(m_image_name);
   m_header_obj = util::header_name(m_image_id);
   m_objmap_name = ObjectMap<>::object_map_name(m_image_id, CEPH_NOSNAP);
+  m_force_non_primary = !non_primary_global_image_id.empty();
+}
+
+template<typename I>
+void CreateRequest<I>::send() {
+  ldout(m_cct, 20) << dendl;
+
+  get_pool_metadata();
+}
+
+template <typename I>
+void CreateRequest<I>::get_pool_metadata() {
+  ldout(m_cct, 20) << "start_key=" << m_last_metadata_key << dendl;
+
+  librados::ObjectReadOperation op;
+  cls_client::metadata_list_start(&op, m_last_metadata_key, MAX_METADATA_ITEMS);
+
+  using klass = CreateRequest<I>;
+  librados::AioCompletion *comp =
+    create_rados_callback<klass, &klass::handle_get_pool_metadata>(this);
+  m_outbl.clear();
+  m_io_ctx.aio_operate(RBD_INFO, comp, &op, &m_outbl);
+  comp->release();
+}
+
+template <typename I>
+void CreateRequest<I>::handle_get_pool_metadata(int r) {
+  ldout(m_cct, 20) << "r=" << r << dendl;
+
+  std::map<std::string, bufferlist> metadata;
+  if (r == 0) {
+    auto it = m_outbl.cbegin();
+    r = cls_client::metadata_list_finish(&it, &metadata);
+  }
+
+  if (r == -EOPNOTSUPP || r == -ENOENT) {
+    ldout(m_cct, 10) << "pool metadata not supported by OSD" << dendl;
+  } else if (r < 0) {
+    lderr(m_cct) << "failed to retrieve pool metadata: " << cpp_strerror(r)
+                 << dendl;
+    complete(r);
+    return;
+  }
+
+  if (!metadata.empty()) {
+    m_metadata.insert(metadata.begin(), metadata.end());
+    m_last_metadata_key = metadata.rbegin()->first;
+    if (boost::starts_with(m_last_metadata_key,
+                           ImageCtx::METADATA_CONF_PREFIX)) {
+      get_pool_metadata();
+      return;
+    }
+  }
 
-  if (image_options.get(RBD_IMAGE_OPTION_FEATURES, &m_features) != 0) {
-    m_features = util::get_rbd_default_features(m_cct);
+  ConfigProxy config{false};
+  std::vector<std::string> keys;
+  m_cct->_conf.get_all_keys(&keys);
+  for (auto &key : keys) {
+    auto it = m_metadata.find(ImageCtx::METADATA_CONF_PREFIX + key);
+    if (it != m_metadata.end()) {
+      std::string value(it->second.c_str(), it->second.length());
+      r = config.set_val(key.c_str(), value);
+      if (r == 0) {
+        continue;
+      }
+      lderr(m_cct) << " failed to set config " << key << " with value "
+                   << it->second.c_str() << ": " << cpp_strerror(r) << dendl;
+    }
+    std::string value;
```

Comment:
Nit: perhaps just use a copy-constructor to initialize a local ```ConfigProxy``` from the existing ```m_cct->_conf``` and then only apply the overrides.

Reply:
@dillaman The copy-constructor is implicitly deleted due to Mutex member.

Reply:
OK, how about a `clone` or `dup` helper method that initializes one config from another?

Reply:
@dillaman updated. I had problems with `dup` implementation due to `const` variables (in `ConfigValues` and `is_daemon`). I have added an explicit copy-constructor, though it looks dangerous to me in general case to copy values without holding the source lock. What do you think?

Reply:
> What do you think?

You could use a static helper method to take the lock to safely copy config within the initializer list.

Reply:
@dillaman sorry, I am not sure I understand what you mean here.

Reply:
Something like:

```
class ConfigProxy {
...
  static ConfigValues get_config_values(const ConfigProxy &config_proxy) {
    Mutex::Locker locker(config_proxy.lock);
    return config_proxy.values;
  }

...
public:
....
    explicit ConfigProxy(const ConfigProxy &config_proxy)
    : values(get_config_values(config_proxy.values)),
      config{values, obs_mgr, config_proxy.config.is_daemon},
      lock{"ConfigProxy", true, false}
  {}
...
};

Reply:
Ah, thanks. I was thinking about this but was not sure about the order, I thought it could be that `locker` is destroyed before the result is returned (copied).

Reply:
@dillaman Updated, thank!

(now I can't understand why I thought it might not work, as it is our usual practice).